src/compiler/checker.ts

```diff
@@ -26595,9 +26602,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             case SyntaxKind.PropertyAccessExpression:
             case SyntaxKind.ElementAccessExpression:
                 const sourcePropertyName = getAccessedPropertyName(source as AccessExpression);
-                const targetPropertyName = isAccessExpression(target) ? getAccessedPropertyName(target) : undefined;
-                return sourcePropertyName !== undefined && targetPropertyName !== undefined && targetPropertyName === sourcePropertyName &&
-                    isMatchingReference((source as AccessExpression).expression, (target as AccessExpression).expression);
+                if (sourcePropertyName !== undefined) {
+                    const targetPropertyName = isAccessExpression(target) ? getAccessedPropertyName(target) : undefined;
+                    if (targetPropertyName !== undefined) {
+                        return targetPropertyName === sourcePropertyName && isMatchingReference((source as AccessExpression).expression, (target as AccessExpression).expression);
+                    }
+                }
+                if (isElementAccessExpression(source) && isElementAccessExpression(target) && isIdentifier(source.argumentExpression) && isIdentifier(target.argumentExpression)) {
+                    const symbol = getResolvedSymbol(source.argumentExpression);
+                    if (symbol === getResolvedSymbol(target.argumentExpression) && (isConstantVariable(symbol) || isParameterOrMutableLocalVariable(symbol) && !isSymbolAssigned(symbol))) {

```

Comment:
Does it make more sense to do `isPastLastAssignment` using a more specific location? How does this PR interact with #56908?

Reply:
Unfortunately not that simple. We would need to know that _all_ narrowing operations for `obj[key]` occur past the last assignment to `key` which would require an additional CFA graph walk. We don't have that issue in #56908 because the only location we care about is the location where the arrow function is created.

Reply:
Makes sense; I really need to prototype the idea I had to make that efficient enough for us to do, but I've lost that brainwave ☹️

Can you add a test in the vein of #56908 in terms of what closures do? Something like: [Playground Link](https://www.staging-typescript.org/play?ts=5.5.0-pr-57847-24#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXyk11QAowAjALnhIEp4BeAPngDccthbr3OBuALAAoYSnTY8SAIwkA1iACe1AM4YYWVAHN6Ab2HwD8CCAzwc5AFbUASuBwxgAHlXqtAGnguNm+AB94aKCIGiDALAzwOgC+gkKGZpYA2vIKALqM8ABEUORgmbHxYHiqbFIZ5hbJiqkFhgD0dfEAegD8+oaEEqR0jCw68EWoJQAe5Ukp6VG0tQYN8fMLzW1xhhVVaRmBIMGooTMDxaasAExjlRP7c4at7QadxCQ9zJEHQ6ajEWsT8FOXjYsA67LKLCYRzACSyk8WAAtlhoDB4BgcAA6NGgoRiIiSRDHHp6FYGYymCoqNTePwBVBBELATbU7a0-YVDLZXL5W6vEqsMqfSx-Jac+54R70Z79QYjM4-aacq6AgE3QkJCz0mm7YD7SVHU58iwCoFC7HdMV9LnvaW-OX-BWLJVRIA) (which does behave properly I think)

Reply:
> Can you add a test in the vein of...

Such a test wouldn't show anything. The new analysis in #56908 only affects references to local variables, not references to properties, so there are no changes across closure boundaries with this PR.